code/game/machinery/shieldgen.dm

```diff
@@ -553,3 +553,644 @@
+				var/obj/structure/emergency_shield/modular/deploying_shield = new(target_tile)
+				deploying_shield.shield_generator = src
+				deployed_shields += deploying_shield
+		calculate_regeneration()
+		active_power_usage += deployed_shields.len * BASE_MACHINE_ACTIVE_CONSUMPTION * 0.1
+		return
+
+	//this code only runs on radius less than 10 and gives us a more accurate circle that is more compatible with decimal values
+	var/list/inside_shield = circle_range_turfs(src, radius - 1)//in the future we might want to apply an effect to the turfs inside the shield
+	if(exterior_only)
+		for(var/turf/target_tile as anything in circle_range_turfs(src, radius))
+			if (!(target_tile in inside_shield) && isspaceturf(target_tile) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
+				var/obj/structure/emergency_shield/modular/deploying_shield = new(target_tile)
+				deploying_shield.shield_generator = src
+				deployed_shields += deploying_shield
+		calculate_regeneration()
+		active_power_usage += deployed_shields.len * BASE_MACHINE_ACTIVE_CONSUMPTION * 0.1
+		return
+
+	for(var/turf/target_tile as anything in circle_range_turfs(src, radius))
+		if (!(target_tile in inside_shield) && isopenturf(target_tile) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
+			var/obj/structure/emergency_shield/modular/deploying_shield = new(target_tile)
+			deploying_shield.shield_generator = src
+			deployed_shields += deploying_shield
+	calculate_regeneration()
+	active_power_usage += deployed_shields.len * BASE_MACHINE_ACTIVE_CONSUMPTION * 0.1
+	return
+
+
+/obj/machinery/modular_shield_gen/Destroy()
+	QDEL_LIST(deployed_shields)
+	active_power_usage = BASE_MACHINE_ACTIVE_CONSUMPTION * 0.5
+	return ..()
+
+/obj/machinery/modular_shield_gen/update_icon_state()
+
+	icon_state = ("gen_[!(machine_stat & NOPOWER)?"[recovering ?"recovering_":"ready_"]":"no_power_"][(panel_open)?"open":"closed"]")
+	return ..()
+
+//ui stuff
+/obj/machinery/modular_shield_gen/ui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Modularshieldgen")
+		ui.open()
+
+/obj/machinery/modular_shield_gen/ui_data(mob/user)
+
+	var/list/data = list()
+	data["max_radius"] = max_radius
+	data["current_radius"] = radius
+	data["max_strength"] = max_strength
+	data["max_regeneration"] = max_regeneration
+	data["current_regeneration"] = current_regeneration
+	data["current_strength"] = stored_strength
+	data["active"] = active
+	data["recovering"] = recovering
+	data["exterior_only"] = exterior_only
+	return data
+
+/obj/machinery/modular_shield_gen/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+	switch(action)
+		if ("set_radius")
+			if (active)
+				return
+			radius = max(1,(text2num(params["new_radius"])))
+		if ("toggle_shields")
+			toggle_shields()
+			return
```

Comment:
```suggestion

```

not useful because there's no more possible logic that we should avoid executing